### PR TITLE
Add new KFServing sample for GCP IAP

### LINKED ISSUE #852: Add KFserving sample for GCP Identity Aware Proxy
/kind feature

**Describe the solution you'd like**
Additional KFServing sample to describe how an InferenceService can be created behind a GCP Identity Aware Proxy (IAP).  The process is somewhat complicated and this will help new users who wish to use KFServing on GCP.

**Anything else you would like to add:**
I'll be doing a pull request against this Issue soon.  Much of the background info was provide by @yuzisun on the kubeflow KFServing slack channel (thanks!)

A summary of what the new sample will include: 
1. **Enable GCP IAP**
1. **Change firewall rules (for private k8s clusters)**
1. **Create the inference service**
1. **Test the private predict endpoint (using port-forwarding)**
1. **Expose the inference service externally using an additional Istio Virtual Service**
1. **Test the external predict endpoint (using iap_request.py)**

### PR DESCRIPTION
**Pull request by owennewo**
**What this PR does / why we need it**:
This PR adds a sample for how to use KFServing behind a GCP Identity Aware Proxy

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #852

**Special notes for your reviewer**:

**Release note**:
```NONE

```

### CODE REVIEW

**Comment by owennewo** [comment]
@yuzisun - a couple of weeks ago you suggested I write up 'kfserving behind GCP IAP'.  Here is is. 
Let me know if there is anything missing or you want anything changed

**Comment by yuzisun** [comment]
This is awesome! thanks @owennewo !

/ok-to-test

File: docs/samples/gcp-iap/README.md [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+```
+kubectl apply -f sklearn-iap-no-authz.yaml
+```
+
+Expected Output
+```
+$ inferenceservice.serving.kubeflow.org/sklearn-iap created
+```
+
+As well as the InferenceService, the kfserving webhook typically creates a Deployment, Replicaset, Pod(s) + Knative Service, Route, Configuration and Revision!
```
Comment (yuzisun) [inline docs/samples/gcp-iap/README.md:60]: might want to rephrase a bit, KFServing controller creates the `Knative Service` and `Knative Service` creates the configuration, revision and route.

File: docs/samples/gcp-iap/README.md [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+We assume that you have created Kubeflow profiles and that [GCP workload identity](https://www.kubeflow.org/docs/gke/authentication/) is configured.
+
+This may not be essential for this example, but is recommended as it maps GCP service-accounts (understood by the IAP) to kubernetes service-accounts (understood by kubernetes and istio RBAC).
+
+### Change firewall rules (for private k8s clusters) 
+If you are running Kubeflow in a [private GKE cluster](https://cloud.google.com/kubernetes-engine/docs/how-to/private-clusters) it will be configured with a restrictive firewall.  
+You will need to ensure that the k8s master node(s) can talk to the following k8s worker node(s) ports: 
+ - 6443 (for cert-manager) 
+ - 8443 (for probes such as kfserving)
```
Comment (yuzisun) [inline docs/samples/gcp-iap/README.md:35]: what's this probe ?

File: docs/samples/gcp-iap/README.md [hunk 3]
```diff
@@ -0,0 +1,137 @@
+## KFServing GCP/IAP Example 
+When using Kubeflow with GCP it is common to use a [GCP Identity Aware Proxy](https://cloud.google.com/iap) (IAP) to manage client authentication to the KFServing endpoints.  The proxy intercepts and authenticates users and passes identity assertion (JWT) to kubernetes service/pods.  Whilst it is also possible to add access control (i.e. programmable or service mesh authorization), this is not described here.
+
+### Prerequisites
+1. Your ~/.kube/config should point to a cluster with [KFServing installed](https://github.com/kubeflow/kfserving/blob/master/docs/DEVELOPER_GUIDE.md#deploy-kfserving).
```
Comment (yuzisun) [inline docs/samples/gcp-iap/README.md:5]: might want to mention the kubeflow installation as well
* You can use [kfctl_gcp_iap.v1.0.2.yaml](https://raw.githubusercontent.com/kubeflow/manifests/v1.0-branch/kfdef/kfctl_gcp_iap.v1.0.2.yaml), editing the repos section in the yaml before deploying to your Kubernetes cluster which contains [knative istio probe fix](https://github.com/kubeflow/manifests/commit/928cf483361730121ac18bc4d0e7a9c129f15ee2).

    ```diff
        repos:
        - name: manifests
    +     uri: https://github.com/kubeflow/manifests/archive/master.tar.gz
    -     uri: https://github.com/kubeflow/manifests/archive/v1.0.2.tar.gz
    -   version: v1.0.2
    ```

File: docs/samples/gcp-iap/make-prediction.sh [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+NAMESPACE='<namespace>'
+INGRESS_DNS=kubeflow.endpoints.${PROJECT}.cloud.goog
+IAP_CLIENT_ID='<project-id>-<random-32-hash>.apps.googleusercontent.com'
+SERVICE_URL=https://${INGRESS_DNS}/kfserving/${NAMESPACE}/${INFERENCE_SERVICE}:predict
+
+# Print out the command that can be used to execute a http prediction request
+echo python iap_request.py $SERVICE_URL ${IAP_CLIENT_ID} --input=${INPUT_PATH}
+
+# Uncomment next line to execute
+# python iap_request.py $SERVICE_URL ${IAP_CLIENT_ID} --input=${INPUT_PATH}
```
Comment (yuzisun) [inline docs/samples/gcp-iap/make-prediction.sh:28]: nit: newline

File: docs/samples/gcp-iap/README.md (hunk 2 above)
Comment (owennewo) [reply docs/samples/gcp-iap/README.md:35]: OK Kfserving probe is the wrong term but the 8433 port is required.
Without it I get a timeout after ~30s
```rror from server (InternalError): error when creating "sklearn-iap-no-authz.yaml": Internal error occurred: failed calling webhook "inferenceservice.kfserving-webhook-server.defaulter": Post https://kfserving-webhook-server-service.kubeflow.svc:443/mutate-inferenceservices?timeout=30s: dial tcp 10.36.1.33:9876: i/o timeout```
I know it mentions 443 but allowing access to 8433 resolves the issue.
I'm going to change it to 'istio readiness probe'.  When I was talking to you about it on slack you mentioned PR #1137.  
I think opening 8443 with private gke cluster is useful for a bunch of stuff tbh e.g. prometheus
https://medium.com/@damianmyerscough/private-gke-cluster-running-cert-manager-prometheus-operator-57b676406d9a

File: docs/samples/gcp-iap/README.md [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+
+**Warning:** The [sklearn-iap-no-authz.yaml](./sklearn-iap-no-authz.yaml) has an annotation that prevents the istio sidecar from being injected and thus disables istio RBAC authorization.  This is unlikely to be suitable for production.
+
+### Test the private predict endpoint (using port-forwarding)
+We can't yet test the inferenceservice predict endpoint externally because GCP IAP only supports path-based routing but KFServing has exposed a host-based routing url.  We can, however, test the private (internal) service using port forwarding.
+
+We can use kubectl port-forward (bypassing IAP) to tunnel to the private service:
+```
+# terminal 1
+kubectl port-forward svc/sklearn-iap-predictor-default-<guid>-private 8080:80
```
[the remark this answers is not in the document]
Comment (owennewo) [reply docs/samples/gcp-iap/README.md]: I've removed section on port-forwarding to test private service

File: docs/samples/gcp-iap/README.md (hunk 2 above)
Comment (owennewo) [reply docs/samples/gcp-iap/README.md]: I'm going to mark this as resolved - feel free to suggest a better name than 'istio readiness probe'

File: docs/samples/gcp-iap/README.md [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+
+We assume that you have created Kubeflow profiles and that [GCP workload identity](https://www.kubeflow.org/docs/gke/authentication/) is configured.
+
+This may not be essential for this example, but is recommended as it maps GCP service-accounts (understood by the IAP) to kubernetes service-accounts (understood by kubernetes and istio RBAC).
+
+### Change firewall rules (for private k8s clusters) 
+If you are running Kubeflow in a [private GKE cluster](https://cloud.google.com/kubernetes-engine/docs/how-to/private-clusters) it will be configured with a restrictive firewall.  
+You will need to ensure that the k8s master node(s) can talk to the following k8s worker node(s) ports: 
+ - 6443 (for cert-manager) 
+ - 8443 (for istio readiness probe)
```
Comment (yuzisun) [inline docs/samples/gcp-iap/README.md:45]:
```suggestion
 - 8443 (for kfserving webhook)
```

File: docs/samples/gcp-iap/README.md [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+
+To create these rules firewall rules, modify and run:
+
+```
+gcloud compute firewall-rules create kubeflow-webhook-probe --allow=tcp:8443 --target-tags kubeflow-worker --direction INGRESS --network default --priority 1000 --source-ranges 172.16.0.0/28
```
Comment (yuzisun) [inline docs/samples/gcp-iap/README.md:50]: tcp:6443?

**Comment by yuzisun** [comment]
@owennewo thanks for this great doc! just a few final nits we are good to go.

File: docs/samples/gcp-iap/README.md (hunk 7 above)
Comment (owennewo) [reply docs/samples/gcp-iap/README.md:50]: Good spot I've changed line 51 (cert-manager) to open port 6443

**Comment by yuzisun** [comment]
Thanks @owennewo for this great doc ! we can follow up with the path based routing virtual service work if you are interested in working on that.
 
/lgtm
/approve

**Pull request merged**